dashboard/app/helpers/lti_helper.rb

```diff
@@ -0,0 +1,119 @@
+
+module LtiHelper
+
+  def windows_complex_password
+    # Take away # to avoid possible &# combo ...
+    # specials = "~!@\$%^&*+=?".chars.to_a
+
+    numbers  = (0..9).to_a
+    [0, 1].each{ |ambiguous_character|
+      numbers.delete ambiguous_character
+    }
+
+    alpha = ('a'..'z').to_a + ('A'..'Z').to_a
+    %w{A E I O U a e i o u l}.each{ |ambiguous_character|
+      alpha.delete ambiguous_character
+    }
+
+    characters = (alpha + numbers)
+    password = Random.new.rand(7..11).times.map{characters.sample}
+
+    # remove special chars for now...
+    #
+    # password << specials.sample unless
+    #   password.join =~ Regexp.new(Regexp.escape(specials.join))
+
+    password << numbers.sample unless
+      password.join =~ Regexp.new(Regexp.escape(numbers.join))
+
+    password.shuffle.join
+  end
+
+  #
+  # ===========================================================
+  #   LTI Signin
+  # ===========================================================
+  #
+  def lti_signin_and_redirect(params, redirect_url)
+    consumer_key = params[:oauth_consumer_key]
+    consumer_secret = OauthCredentials.get_secret(consumer_key)
+    if consumer_secret.blank?
+      redirect_to new_user_registration_path,
+        alert: "Sorry, invalid consumer_key [#{consumer_key}]"
+      return
+    end
+
+    provider = IMS::LTI::ToolProvider.new(consumer_key, consumer_secret, params)
+    if provider.valid_request?(request)
+      delta = Time.now.to_i - params[:oauth_timestamp].to_i
+
+      # 5 minute window
+      if delta.abs > 60 * 5
```

Comment:
Nitpick - If the caller doesn't provide a timestamp, I think this will say that the launch request is outside the time window (nil.to_i is 0). Might as well provide a more detailed message here